src/utils/userLocation.ts

```diff
@@ -0,0 +1,83 @@
+import firebase, { UserInfo } from "firebase/app";
+
+import { AnyRoom } from "types/rooms";
+
+import { updateUserProfile } from "pages/Account/helpers";
+import { useInterval } from "hooks/useInterval";
+
+import { getCurrentTimeInMilliseconds } from "./time";
+import { openRoomUrl, openUrl, venueInsideUrl } from "./url";
+
+const LOCATION_INCREMENT_SECONDS = 10;
+const LOCATION_INCREMENT_MS = LOCATION_INCREMENT_SECONDS * 1000;
+
+type LocationData = Record<string, number>;
+
+export const updateLocationData = (
+  userId: string,
+  newLocationData: LocationData = {}
+) => {
+  updateUserProfile(userId, {
+    lastSeenAt: getCurrentTimeInMilliseconds(),
+    lastSeenIn: newLocationData,
+  });
+};
+
+export interface TrackLocationProps {
+  userId: string;
+  locationName: string;
+  lastSeenIn?: LocationData;
+}
+
+export const trackLocationEntered = ({
+  userId,
+  locationName,
+}: TrackLocationProps) => {
+  updateLocationData(userId, {
+    [locationName]: getCurrentTimeInMilliseconds(),
+  });
+};
+
+export const enterVenue = (venueId: string) => openUrl(venueInsideUrl(venueId));
+
+type EnterExternalRoomProps = {
+  userId: string;
+  room: AnyRoom;
+};
+
+export const enterExternalRoom = ({ userId, room }: EnterExternalRoomProps) => {
+  trackLocationEntered({
+    userId,
+    locationName: room.url,
```

Comment:
Why `room.url`? I don't think this is correct? Or at the very least I believe, this is a change in logic from what we were previously doing.

It should be using the `venueName/roomName` slug.

Reply:
Yes, it is a change of logic. I don't understand the reasoning behind using `venueName/roomName` slug tbh. This slug amplifies the entry point rather than the location itself, which is not what we want, I guess

Reply:
One reason is that that is how it has always been, and we haven't designed or discussed changing it. There is nothing in this PR that handles migrating existing data. None of the related scripts/reporting/etc are updated for it. We can't just decide to change a core feature of the platform without raising it with the team, and discussing the potential impacts. And we DEFINITELY should not be changing that haphazardly hidden among a giant PR that doesn't mention anything about changing it. This is the sort of things that I raised in retro yesterday where there needs to be more up front planning/discussion/etc.